### DISCUSSION TITLE
Export object on bus name of GApplication

### DISCUSSION THREAD

**Discussion by swsnr** [gtk-rs/gtk-rs-core#1553] [Q&A] [answered]
I'm not too familiar with neither the whole GIO application nor DBus, so please excuse me if I'm asking a stupid question.

I understand that a GIO application automatically claims its application ID as bus name on the session bus.  At least, I'm seeing the name claimed just by launching my application, without having done anything towards DBus registration.

Now, I'd like to export an object under that bus name.  <https://github.com/gtk-rs/gtk-rs-core/issues/1248> suggests I should use zbus, but with the bus name already claimed by the `GApplication` instance and its `GDBusConnection`, how would I do that?  There does not seem to be any way to get zbus work over a `GDBusConnection`, as far as I can see?

`GDBusConnection` in turn has [`register_object`](https://gtk-rs.org/gtk-rs-core/stable/0.20/docs/gio/struct.DBusConnection.html#method.register_object) but it's not documented; its return type, which presumably provides some kind of builder interface, doesn't even appear in the API docs at all.

So, how do I get my own object to appear under my app's bus name?

**Comment by sdroege** [comment, marked as answer]
`GApplication` registers your application's name on the bus. If you want to use zbus you need to select a different name for that and have a separate dbus connection.

For using GDbus, the API works basically the same as in C. In other words, not very convenient until someone puts some effort into making it nicer to use (e.g. via zbus-style macros). https://github.com/gtk-rs/gtk-rs-core/blob/main/examples/gio_dbus_register_object/main.rs is an example that probably answers your question.

**Reply by swsnr** [reply]
Perfect, thanks, that's the example I was looking for. I'll try to see whether I can make that work for my use case 🙂

**Reply by swsnr** [reply]
@sdroege  So, I've now used the GDBus API to [implement a search provider](https://github.com/swsnr/turnon/pull/11), and I think it's pretty okay.  Not much worse than zbus, in my opinion.

The serialization code mostly comes for free, since variant conversions exists for basic types, and can be derived from structs.  The most striking drawback is that method names must be matched manually, but beyond that the difference is mostly syntactical.

**Reply by bilelmoussaoui** [reply]
there is a zbus crate https://crates.io/crates/search-provider

**Reply by swsnr** [reply]
I know, but it's no good for me.  The license is a show-stopper already, and I'm not happy about some of its public API.  And as any zbus-based approach it wouldn't let me expose the search provider under the app's bus name.

**Reply by bilelmoussaoui** [reply]
>  The license is a show-stopper already

We can fix that, it should probably be MIT.

> let me expose the search provider under the app's bus name.

Why is that a requirement? None of the apps expose their search provider under the same bus name.

**Reply by sdroege** [reply]
@swsnr If you think some more examples could be useful here or in the gtk4-rs repo, feel free to submit a PR or two :)

**Reply by swsnr** [reply]
@sdroege I'll see if I find time to submit an example with dispatching and deserialization.  The initial example was very helpful, because the builder is entirely undocumented, but beyond that I managed pretty well :slightly_smiling_face: 

@bilelmoussaoui I've seen both :slightly_smiling_face: It's not a requirement for the search provider for sure, but the search provider was just my test case for exposing an object with gdbus, since it has a fairly large API.   For my own DBus API it matters, if only for aesthetics in `busctl` calls :slightly_smiling_face: 

And now that I've used gdbus and found that it's quite not as bad as suggested in the other ticket, I see no reason to buy the whole heavy dependency chain of zbus.  As said in the ticket it'd probably have doubled the `cargo tree` of my application, and perhaps compilation times too :slightly_smiling_face: My laptop's not that powerful; writing the gdbus code was fun, waiting for zbus to compile isn't so much :wink:

**Reply by swsnr** [reply]
@sdroege I opened https://github.com/gtk-rs/gtk-rs-core/pull/1558 with two small quality of life helpers, and updated the gdbus example with an enum-based call dispatching and parameter handling using auto-derived variant structs.

It now reflects how I'm using gdbus in my app, and for simple interfaces it's reasonably convenient this way.
